creusot-contracts/src/ptr_own.rs

```diff
@@ -0,0 +1,107 @@
+#[cfg(creusot)]
+use crate::util::SizedW;
+use crate::*;
+
+/// Raw pointer whose ownership is tracked by a ghost [PtrOwn].
+pub type RawPtr<T> = *const T;
+
+/// Token that represents the ownership of a memory cell. A [PtrOwn] value only
+/// exists in the ghost world, but can be used in combination with a
+/// corresponding [RawPtr] to access and modify memory.
+///
+/// A warning regarding memory leaks: dropping a `GhostBox<PtrOwn<T>>` (we only
+/// ever handle ghost [PtrOwn] values) cannot deallocate the memory
+/// corresponding to the pointer because it is a ghost value. One must thus
+/// remember to explicitly call [drop] in order to free the memory tracked by a
+/// [PtrOwn] token.
+#[trusted]
+pub struct PtrOwn<T: ?Sized>(std::marker::PhantomData<T>);
+
+impl<T: ?Sized> PtrOwn<T> {
+    #[trusted]
+    #[logic]
+    pub fn ptr(&self) -> RawPtr<T> {
+        dead
+    }
+
+    #[trusted]
+    #[logic]
+    pub fn val(&self) -> SizedW<T> {
+        dead
+    }
+}
+
+impl<T: ?Sized> Invariant for PtrOwn<T> {
+    #[predicate(prophetic)]
+    #[open]
+    #[creusot::trusted_ignore_structural_inv]
+    #[creusot::trusted_is_tyinv_trivial_if_param_trivial]
+    fn invariant(self) -> bool {
+        pearlite! { !self.ptr().is_null_logic() && inv(self.val()) }
+    }
+}
+
+/// Creates a new [PtrOwn] and associated [RawPtr] by allocating a new memory
+/// cell initialized with `v`.
+#[ensures(result.1.ptr() == result.0 && *result.1.val() == v)]
+pub fn new<T>(v: T) -> (RawPtr<T>, GhostBox<PtrOwn<T>>) {
+    from_box(Box::new(v))
+}
+
+/// Creates a ghost [PtrOwn] and associated [RawPtr] from an existing [Box].
+#[trusted]
+#[ensures(result.1.ptr() == result.0 && *result.1.val() == *val)]
+pub fn from_box<T: ?Sized>(val: Box<T>) -> (RawPtr<T>, GhostBox<PtrOwn<T>>) {
+    assert!(core::mem::size_of_val::<T>(&*val) > 0, "PtrOwn doesn't support ZSTs");
+    (Box::into_raw(val), GhostBox::conjure())
+}
+
+/// Immutably borrows the underlying `T`.
+#[trusted]
+#[requires(ptr == own.ptr())]
+#[ensures(*result == *own.val())]
+#[allow(unused_variables)]
+pub fn as_ref<T: ?Sized>(ptr: RawPtr<T>, own: GhostBox<&PtrOwn<T>>) -> &T {
+    unsafe { &*ptr }
+}
+
+/// Mutably borrows the underlying `T`.
+#[trusted]
+#[allow(unused_variables)]
+#[requires(ptr == own.ptr())]
+#[ensures(*result == *own.val())]
+// TODO: why is .inner_logic() needed instead of *?
```

Comment:
@arnaudgolfouse do you know what's happening here? (below)

Reply:
Yeah, I give more details in #1257 ... The bottom line is that using `*` was a hack anyways, and it does not work in this case 😢